garage/tf/algos/npo.py

```diff
@@ -256,32 +256,56 @@ def _build_policy_loss(self, i):
                     m = tf.reduce_min(adv_valid)
                     adv_valid = (adv_valid - m) + eps
 
-            policy_dist_info_flat = self.policy.dist_info_sym(
-                i.flat.obs_var,
-                i.flat.policy_state_info_vars,
-                name="policy_dist_info_flat")
-            policy_dist_info_valid = filter_valids_dict(
-                policy_dist_info_flat,
-                i.flat.valid_var,
-                name="policy_dist_info_valid")
+            if self.policy.recurrent:
+                policy_dist_info = self.policy.dist_info_sym(
+                    i.obs_var,
+                    i.policy_state_info_vars,
+                    name="policy_dist_info")
+            else:
+                policy_dist_info_flat = self.policy.dist_info_sym(
```

Comment:
I wasn't aware we could compute the dense versions of `dist_info`, `kl`, and `lr`.

Please verify that this actually works.

If it works, please remove the flat/valid views from NPO and just flatten/filter valids everything right before the final loss function, which is less code.

Reply:
The recurrent policy's `dist_info` requires a shape of `(trajectory, step, data_flat_dim)`, but the nonrecurrent policy's `dist_info` requires an input shape of `(trajectory x step, data_flat_dim)`. I think we still need flat/valid views for nonrecurrent policies.

Reply:
Alright thanks for checking! We can fix this when we remove garage.distributions